WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java

```diff
@@ -0,0 +1,397 @@
+    }
+
+    protected void next(String email) {
+        if (!checkNetworkConnectivity()) {
+            return;
+        }
+
+        if (isValidEmail(email)) {
+            showEmailCheckProgressDialog();
+            mDispatcher.dispatch(AccountActionBuilder.newIsAvailableEmailAction(email));
+        } else {
+            showEmailError(R.string.email_invalid);
+        }
+    }
+
+    private String getEmail() {
+        return EditTextUtils.getText(mEmailEditText).trim();
+    }
+
+    private boolean isValidEmail(String email) {
+        Pattern emailRegExPattern = Patterns.EMAIL_ADDRESS;
+        Matcher matcher = emailRegExPattern.matcher(email);
+
+        return matcher.find() && email.length() <= MAX_EMAIL_LENGTH;
+    }
+
+    private final OnClickListener mNextClickListener = new OnClickListener() {
+        @Override
+        public void onClick(View v) {
+            next(getEmail());
+        }
+    };
+
+    @Override
+    public void afterTextChanged(Editable s) {
+    }
+
+    @Override
+    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+    }
+
+    @Override
+    public void onTextChanged(CharSequence s, int start, int before, int count) {
+        UpdateNextButton();
+        mEmailEditTextLayout.setError(null);
+    }
+
+    private void UpdateNextButton() {
+        if (fieldsFilled()) {
+            mNextButton.setEnabled(true);
+        } else {
+            mNextButton.setEnabled(false);
+        }
+    }
+
+    private boolean fieldsFilled() {
+        return EditTextUtils.getText(mEmailEditText).trim().length() > 0;
+    }
+
+    protected boolean isUserDataValid() {
+        final String email = EditTextUtils.getText(mEmailEditText).trim();
+        boolean retValue = true;
+
+        if (TextUtils.isEmpty(email)) {
+            showEmailError(R.string.required_field);
+            retValue = false;
+        }
+
+        return retValue;
+    }
+
+    private void showEmailError(int messageId) {
+        mEmailEditTextLayout.setError(getString(messageId));
+        mEmailEditText.post(new Runnable() {
+            @Override
+            public void run() {
+                mEmailEditText.requestFocus();
+            }
+        });
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        mDispatcher.register(this);
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+        mDispatcher.unregister(this);
+    }
+
+    private void showEmailCheckProgressDialog() {
+        startProgress(getActivity().getString(R.string.checking_email));
+    }
+
+    @Override
+    protected void startProgress(String message) {
+        mNextButton.setEnabled(false);
+        mProgressDialog = ProgressDialog.show(getActivity(), "", message, true, true,
+                new DialogInterface.OnCancelListener() {
+            @Override
+            public void onCancel(DialogInterface dialog) {
+                UpdateNextButton();
```

Comment:
I've already commented on this a bit in `onAvailabilityChecked` but I just wanted to say that I don't think we should use the `dialog`'s state as a flag. There is no guarantee that another developer will come in and change the behavior of the dialog, not realizing how it was intended to use. So, my suggestion would be if we need a flag here (not saying we do, because it depends on how we'll handle the duplicate requests) we should just use a `boolean` variable.

As a rule of thumb, I personally like to avoid making things do more than _one_ thing. An initially clever solution can sometimes become a hassle when multiple developers work on the same project. Just my 2 cents.